common/transform/src/main/java/ai/h2o/mojos/deploy/common/transform/ShapleyScoreException.java

```diff
@@ -0,0 +1,7 @@
+package ai.h2o.mojos.deploy.common.transform;
+
+public class ShapleyScoreException extends Exception {
+  public ShapleyScoreException(String message) {
+    super(message);
+  }
```

Comment:
Is there a future purpose for this custom exception type?  Looks like the exception travels just a single frame (e.g., `MojoScorer` up to `ModelsApiController`), and is unpacked immediately.

Reply:
Good question. I think the intent originally was to embed the error without throwing the error (400) so whoops on my part. @orendain wdyt about that though? If shapley scoring fails for `/model/score` should we return the score and a message saying shapley failed? or should we just fail outright?

Reply:
> If shapley scoring fails for /model/score should we return the score and a message saying shapley failed? or should we just fail outright?

Hmmm, lots of pros/cons floating around in my head ... For me, it helps me to think in terms of requirements and best practices:

- Generally, there's no guarantee that scoring and contributions use the same backend (in fact, this scorer has separate pipelines for each).  It's possible for 100% contribution failure while scoring succeeds 100% of the time.  Defending scoring endpoint against contribution failure makes sense.  (+1 for not failing b/c contributions)
- If we apply the single responsibility principle loosely, `/model/score` should only be focused on scoring, and one could argue that contributions are at most a bonus (and not required as they have their own endpoint).  (+1 for not failing b/c contributions)
- If you do return messages:  If contribution fails (say, consistently), adding a message w/o it being expected might mean contribution concerns hijack scoring packet size.  Especially since error messages could easily make up 50%+ of response packet for single-row scores.  (+1 to [user explicitly requesting messages](https://github.com/h2oai/dai-deployment-templates/issues/165))
- If a user explicitly asks for contributions as part of `/model/score`, and you decide to return a message instead of failing outright, requiring users to "opt-in" to warning messages might actually be a form of failing silently.  That is, unless the user explicitly asked for messages, they won't get any, even when they're explicitly asked for contributions to be returned and would expect to be notified of failure. (+1 to interpreting users asking for contributions as opt-in for warning-messages).
- If you do return messages:  Modifying API spec to add message-related fields will change expectation of other scorers.  Consider that other scorers updating to the latest API spec means they should implement any new features, unless you mark them as implementation-optional in the API.

Just some thoughts in case it helps to think about the problem - no hard opinion floating up there other than there should definitely be _some_ mechanism by which users are made aware of issues.

Reply:
removed, unneeded.

Reply:
Hmmm so this is a tricky one....
1. I like the idea of not failing `/model/score` because of contributions failing.
2. I like the idea of users opting into error messages but in reality the only place where messages would be "opt-in" would be for `/model/score` where shapley is requested.
3. I would argue that technically this feature is already in the python scorer api, and that due to limitations in swagger this was never implemented.

So there maybe is a better way to do this by implementing a proper exception handler for the controller. I'm going to take another shot at this later. It just really bugs me that we literally have no feedback to the user on errors.